pkg/reconciler/trigger/trigger.go

```diff
@@ -61,43 +68,51 @@ type TriggerReconciler struct {
 	uriResolver        *resolver.URIResolver
 
 	// projectID is used as the GCP project ID when present, skipping the
-	// metadata server check. Used by tests.
+	// metadata server check.
 	projectID string
 
 	// pubsubClient is used as the Pubsub client when present.
 	pubsubClient *pubsub.Client
 }
 
 // Check that TriggerReconciler implements Interface
-var _ triggerreconciler.Interface = (*TriggerReconciler)(nil)
-var _ triggerreconciler.Finalizer = (*TriggerReconciler)(nil)
-
-func (r *TriggerReconciler) ReconcileKind(ctx context.Context, t *brokerv1beta1.Trigger) pkgreconciler.Event {
-	b := brokerFromContext(ctx)
-	if b == nil {
-		// Assume the Broker has been deleted or doesn't exist yet. Create a
-		// Broker object with the expected name and namespace to
-		// reconcile with.
-		// TODO move this to resources package?
-		b = &brokerv1beta1.Broker{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      t.Spec.Broker,
-				Namespace: t.Namespace,
-			},
-			//TODO is this needed?
-			Status: brokerv1beta1.BrokerStatus{},
-		}
-		b.Status.InitializeConditions()
-		//return fmt.Errorf("Couldn't fetch Broker from context")
+var _ triggerreconciler.Interface = (*Reconciler)(nil)
+var _ triggerreconciler.Finalizer = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, t *brokerv1beta1.Trigger) pkgreconciler.Event {
+	b, err := r.brokerLister.Brokers(t.Namespace).Get(t.Spec.Broker)
+
+	// If there is an error getting the Broker or the Broker is deleted, we mark the Trigger status
+	// regardless of whether the Trigger belongs to this controller or not. It doesn't hurt.
+	if err != nil && !apierrs.IsNotFound(err) {
+		msg := fmt.Sprintf("Unknown error getting broker: %v", err)
+		logging.FromContext(ctx).Error(msg)
+		t.Status.MarkBrokerUnknown("UnknownBroker", msg)
+		return err
 	}
-	t.Status.InitializeConditions()
-
-	if b.DeletionTimestamp != nil || b.UID == "" {
+	// If the broker has been or is being deleted, we clean up resources created by this controller
+	// for the given trigger.
+	if apierrs.IsNotFound(err) || !b.GetDeletionTimestamp().IsZero() {
 		t.Status.MarkBrokerFailed("BrokerDoesNotExist", "Broker %q does not exist", t.Spec.Broker)
-	} else {
-		t.Status.PropagateBrokerStatus(&b.Status)
+		return r.FinalizeKind(ctx, t)
 	}
 
+	if !filterBroker(b) {
+		// Call Finalizer anyway in case the Trigger still holds GCP Broker related resources.
+		// If a Trigger used to point to a GCP Broker but now has a Broker with a different brokerclass,
+		// we should clean up resources related to GCP Broker.
+		return r.FinalizeKind(ctx, t)
```

Comment:
If a trigger is created for a broker that doesn't have this broker class, will the `googlecloud` finalizer be added anyway? I think it's automatically added in genreconciler when FinalizeKind is present, before Reconcile is even called. So most likely every Trigger will have the `googlecloud` finalizer even if they use another broker class.

I think that means that every trigger with a different broker class will be finalized on every update. The unit test of a trigger with a not found broker seems to confirm this (because it finalizes that trigger). Might want to confirm with a real-world test.

If that behavior is confirmed then we should disable the automatic finalizer logic in genreconciler (by renaming FinalizeKind). Then we can add the finalizer only to Triggers that point to a googlecloud broker.

Reply:
Your understanding is correct. I forgot to add a TODO here to explain the finalizer scenario and future improvements.

> I think that means that every trigger with a different broker class will be finalized on every update.

Correct. This is unnecessary and it will also cause read operation on the topic and pull subscription.

>  we should disable the automatic finalizer logic in genreconciler (by renaming FinalizeKind).

Unfortunately if we do this, the FinalizeKind logic won't be called when the trigger object is deleted. The genreconciler will do nothing because it thinks there is no finalizer to call if the reconciler doesn't implement FinalizeKind.

I have thought about this and came up with a few options:
1. The current behavior. Pros: least effort; Cons: unnecessary FinalizeKind calls for non-GCP broker. 
1. Use a custom controller instead of genreconciler. This essentially means we need to copy most of the  genreconciler boil plate code and make some modifications.
1. Upstream genreconciler changes to make it more flexible, specifically, add a `FilterKind` interface to let the genreconciler skip a trigger if it doesn't point to a GCP broker. I have proposed this interface before https://github.com/knative/pkg/issues/1149

Reply:
I personally prefer Option 3. If this sounds good to you, I will add a TODO for that.

Reply:
Dumb question: why would we reconcile such a trigger (trigger not pointing a gcp broker) in the first place? Broker in trigger spec is immutable.

Also for option 2, there might not be that much boilerplate code from *just* the FinalizeKind. We can still use the majority of the genreconciler.

Reply:
> why would we reconcile such a trigger (trigger not pointing a gcp broker) in the first place? Broker in trigger spec is immutable.

Consider this scenario. A trigger was pointed to a gcp broker. Then the broker was updated to use a different brokerclass. Now we want to tear down trigger pubsub resources. We don't know if users will actually do this. But in theory they can.

We could do option 2 but I really don't like the trigger controller to be "special". 
I don't think it's very high  priority because: 1. users are likely to only use gcp broker. 2. even if they use a different broker, this results in a couple of unnecessary calls to pubsub. not great but doesn't really hurt too much. 
For the above reasons I prefer to have a clean solution.

Reply:
I don't think you can change the broker class. https://github.com/knative/eventing/blob/master/pkg/apis/eventing/v1beta1/broker_validation.go#L58-L59

Reply:
A broker could be deleted and replaced with a broker with a different class. From the trigger controller's perspective, this will be the same as a broker changing class.

Reply:
Right it's immutable. But you can still delete a broker and recreate it with a different brokerclass.

Reply:
It's fine to defer this for later. I'd love to end up at option 3, but realistically I'd probably attempt option 2 first, then move the new code upstream later. It will be easier to justify as an upstream option once we have a demonstrated solution.

Reply:
SG. Created an issue to follow up https://github.com/google/knative-gcp/issues/1164

Reply:
Nothing blocking. Just an idea: if we can (or already?) label the trigger with the initial broker class, then even if the broker got recreated with a different class we can still handle that deterministically, right?